fdbbackup/backup.actor.cpp

```diff
@@ -2744,36 +2769,90 @@ ACTOR Future<Void> queryBackup(const char* name,
+			                       errorMessage =
+			                           format("the specified restorable version %lld is not valid", restoreVersion));
+			return Void();
+		}
+
+		if (snapshotVersion != invalidVersion) {
+			Optional<RestorableFileSet> fileSet = wait(bc->getRestoreSet(snapshotVersion, keyRangesFilter));
+			if (fileSet.present()) {
+				result["snapshot_version"] = fileSet.get().targetVersion;
+				for (const auto& rangeFile : fileSet.get().ranges) {
+					JsonBuilderObject object;
+					object["file_name"] = rangeFile.fileName;
+					object["file_size"] = rangeFile.fileSize;
+					object["version"] = rangeFile.version;
+					object["key_range"] = fileSet.get().keyRanges.count(rangeFile.fileName) == 0
+					                          ? "none"
+					                          : fileSet.get().keyRanges.at(rangeFile.fileName).toString();
+					rangeFilesJson.push_back(object);
+					totalRangeFilesSize += rangeFile.fileSize;
+				}
+				for (const auto& log : fileSet.get().logs) {
+					JsonBuilderObject object;
+					object["file_name"] = log.fileName;
+					object["file_size"] = log.fileSize;
+					object["begin_version"] = log.beginVersion;
+					object["end_version"] = log.endVersion;
+					logFilesJson.push_back(object);
+					totalLogFilesSize += log.fileSize;
+				}
+
+				snapshotVersion = fileSet.get().targetVersion;
+
+				TraceEvent("BackupQueryReceivedRestorableFilesSetFromSnapshot")
+				    .detail("SnapshotVersion", snapshotVersion)
+				    .detail("DestinationContainer", destinationContainer)
+				    .detail("KeyRangesFilter", printable(keyRangesFilter))
+				    .detail("ActualRestoreVersion", fileSet.get().targetVersion)
+				    .detail("NumRangeFiles", fileSet.get().ranges.size())
+				    .detail("NumLogFiles", fileSet.get().logs.size())
+				    .detail("RangeFilesBytes", totalRangeFilesSize)
+				    .detail("LogFilesBytes", totalLogFilesSize);
 			} else {
 				reportBackupQueryError(
 				    operationId,
 				    result,
-				    errorMessage = format("the backup for the specified key ranges is not restorable to any version"));
+				    format("no restorable files set found for specified key ranges from snapshotVersion %lld",
+				           snapshotVersion));
+				return Void();
 			}
 		}
 
-		if (restoreVersion < 0 && restoreVersion != latestVersion) {
-			reportBackupQueryError(operationId,
-			                       result,
-			                       errorMessage =
-			                           format("the specified restorable version %lld is not valid", restoreVersion));
-			return Void();
+		state Optional<RestorableFileSet> fileSet;
+		if (snapshotVersion == invalidVersion) {
+			// Using the latest snapshot.
+			wait(store(fileSet, bc->getRestoreSet(restoreVersion, keyRangesFilter)));
+		} else {
+			// We only need to know all the log files from snapshotVersion to restoreVersion.
+			wait(store(fileSet, bc->getRestoreSet(restoreVersion, keyRangesFilter, /*logOnly=*/true, snapshotVersion)));
```

Comment:
how different is this from 2801? can we consolidate these two calls?

Reply:
the 2801 code block is to calculate backup data "up to the user-configured snapshot version". 
this code back adds up the backup data from snapshot version to restorable version.

Reply:
virtual Future<Optional<RestorableFileSet>> getRestoreSet(Version targetVersion,
	                                                          VectorRef<KeyRangeRef> keyRangesFilter = {},
	                                                          bool logsOnly = false,
	                                                          Version beginVersion = -1) = 0;